### PR TITLE
Add `onStartupFinished` plugin activation event

### PR DESCRIPTION
**Pull request by vinokurig**
#### What it does

Add `onStartupFinished` plugin activation event

fixes https://github.com/eclipse-theia/theia/issues/8488
#### How to test

1. Clone any Theia plugin e.g https://github.com/eclipse/che-theia-samples/tree/master/samples/hello-world-backend-plugin
2. Change the activation event in its `package.json` from `*` to `onStartupFinished`
3. Start the plugin.

#### Review checklist

- [ ] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/plugin-ext/src/hosted/browser/hosted-plugin.ts [hunk 1]
```diff
@@ -217,6 +217,7 @@ export class HostedPluginSupport {
                 };
             }
         });
+        this.activateByEvent('onStartupFinished');
```
Comment (tsmaeder) [inline packages/plugin-ext/src/hosted/browser/hosted-plugin.ts:220]: I was wondering why the event should be sent at this location and not another. What parts of the system need to be ready in order for startup to be finished? Not saying it's the wrong place, I just want to understand your thinking.

**Review by azatsarynnyy — COMMENTED** [review]
I've verified it according to "How to test" steps. Works well, but I have some concerns about the activation order.

Why not emit it after a star(*) activation https://github.com/eclipse-theia/theia/blob/d974eb0be2801fd36cbfa0ffbe62cdc71246a0c6/packages/plugin-ext/src/plugin/plugin-manager.ts#L219
as it's in VSCode
https://github.com/microsoft/vscode/blob/0e297b112830bd5be9cf51a848d9314c3887ec42/src/vs/workbench/api/common/extHostExtensionService.ts#L475

Isn't it too early emitting `onStartupFinished` during hosted-plugin initialization?
Won't it happen eager than a star(*) activation?

**Review by azatsarynnyy — COMMENTED** [review]
Or another variant. As [`HostedPluginSupport` is `FrontendApplicationContribution`](https://github.com/eclipse-theia/theia/blob/d3f1912ffba9b1550e1a09677b422bce5cf3df74/packages/plugin-ext/src/main/browser/plugin-ext-frontend-module.ts#L99-L103), wouldn't it better to emit the event on some of the `FrontendApplicationContribution`'s callbacks, e.g. [`onStart`](https://github.com/eclipse-theia/theia/blob/d3f1912ffba9b1550e1a09677b422bce5cf3df74/packages/core/src/browser/frontend-application.ts#L45-L49) or [`onDidInitializeLayout`](https://github.com/eclipse-theia/theia/blob/d3f1912ffba9b1550e1a09677b422bce5cf3df74/packages/core/src/browser/frontend-application.ts#L72-L75)?

**Comment by vinokurig** [comment]
@tsmaeder @azatsarynnyy I think I've found a better place for this event :). Now it is called on application's `ready` state which suites the [event's documentation](https://code.visualstudio.com/api/references/activation-events#onStartupFinished).

**Review by azatsarynnyy — APPROVED** [review]
I believe emitting on `appState.reachedState('ready')` looks late enough as for `onStartupFinished`.

**Comment by akosyakov** [comment]
Did you check VS Code source code to learn the real semantic of this event and align it?

**Comment by vinokurig** [comment]
@akosyakov 
vscode handles this event after the `*` events and the `workspaceContains` events: https://github.com/microsoft/vscode/blob/7e0c1aa76e4ce162679409828af17535d50a4424/src/vs/workbench/api/common/extHostExtensionService.ts#L472-L476
Do you think it is better to handle this event after the `workspaceContains` events here? https://github.com/eclipse-theia/theia/blob/8bf15dc1d9b186b11c1f954b4c51500e611d3072/packages/plugin-ext/src/hosted/browser/hosted-plugin.ts#L415-L422

**Comment by akosyakov** [comment]
yes, i think it should mean the same

**Comment by vinokurig** [comment]
@akosyakov Done

**Comment by jeluard** [comment]
@vinokurig I can see the following error message with latest master:

```js
root ERROR [hosted-plugin: 93564] Unsupported activation events: onStartupFinished, please open an issue: https://github.com/eclipse-theia/theia/issues/new
```

Is there something specific to do to avoid that?

**Comment by vinokurig** [comment]
@jeluard Thank you for reporting the issue, the PR above fixes this problem.

**Pull request merged**